---
title: ""
author: "Tim Vigers & Laura Pyle"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_depth: 5
    toc_float: yes
---

```{r setup, include=FALSE}
library(tidyverse)
library(Seurat) # need these for Differential expression testing
library(SeuratObject)
library(data.table)
library(broom)
library(pROC)
library(caTools)
library(nlme)
library(lme4)
library(caret)
library(tidymodels)
library(doParallel)
library(str2str)
library(DataExplorer)
library(knitr)
library(ranger)
# update the dfs as change (changes in first chunk of version 3 on the Lambda)
#perc_express_cluster <- read.csv("S:/Laura/BDC/Projects/Howard Davidson/perc_express_cluster_TEST.csv")

# check donor cluster level
perc_express_donorcluster  <- read.csv("S:/Laura/BDC/Projects/Howard Davidson/perc_express_donorcluster_TEST.csv")
#rownames(df2) = colnames(perc_express_donorcluster)

# data
home_dir = "S:/Laura/BDC/Projects/Howard Davidson/R01/Data_Raw/Davidson pilot data/"
knitr::opts_knit$set(root.dir = home_dir)
#setwd("S:/Laura/BDC/Projects/Howard Davidson/R01/Data_Raw/Davidson pilot data/")
df2 = readRDS(paste0(home_dir, "OneDrive_1_6-28-2023/normalized.rds"))

```

```{r fun, include = FALSE}
gene_list = function(df, cutoff){
  # df is one cluster w/ all donors at perc_expr level
  dft = as.data.frame(t(df)[-1,])
  donors  = colnames(dft)
  donorgenes = list()
  for(i in donors){
    nam = dft %>% select(i)
    donor_vec = nam %>% filter(.[[1]] > cutoff)  %>% rownames()
    donorgenes[[i]] = donor_vec
  }
  return(donorgenes)
}

# Tims elnet
source("D:/Repositories/shared-resources/Machine Learning/Tim - ElasticNet CV/easy_elasticnet.R")
#source("C:/Repositories/shared-resources/Machine Learning/Tim - ElasticNet CV/easy_elasticnet.R")
```

```{r col clusters, include = FALSE}
# 22.3 for 505 
# Cluster 0
c_0 = perc_express_donorcluster %>% filter(cluster == 0) %>% select(-cluster)
a = gene_list(c_0, 22.3) 
cluster0_genes = Reduce(intersect, a) 


# Cluster 1
c_1 = perc_express_donorcluster %>% filter(cluster == 1) %>% select(-cluster)
a = gene_list(c_1, 28) # 28 for 502
cluster1_genes = Reduce(intersect, a)

# Cluster 2
c_2 = perc_express_donorcluster %>% filter(cluster == 2) %>% select(-cluster)
a = gene_list(c_2, 26)
cluster2_genes = Reduce(intersect, a) #26% for 500 exactly hot dog

# Cluster 3
c_3 = perc_express_donorcluster %>% filter(cluster == 3) %>% select(-cluster)
a = gene_list(c_3, 28)
cluster3_genes = Reduce(intersect, a)
# Cluster 4
c_4 = perc_express_donorcluster %>% filter(cluster == 4) %>% select(-cluster)
a = gene_list(c_4, 31.6)
cluster4_genes = Reduce(intersect, a)
# Cluster 5
c_5 = perc_express_donorcluster %>% filter(cluster == 5) %>% select(-cluster)
a = gene_list(c_5, 29.4)
cluster5_genes = Reduce(intersect, a)
# Cluster 6
c_6 = perc_express_donorcluster %>% filter(cluster == 6) %>% select(-cluster)
a = gene_list(c_6, 21)
cluster6_genes = Reduce(intersect, a)
# Cluster 7
c_7 = perc_express_donorcluster %>% filter(cluster == 7) %>% select(-cluster)
a = gene_list(c_7,26)
cluster7_genes = Reduce(intersect, a)
# Cluster 8
c_8 = perc_express_donorcluster %>% filter(cluster == 8) %>% select(-cluster)
a = gene_list(c_8, 26.3)
cluster8_genes = Reduce(intersect, a)
# Cluster 9
c_9 = perc_express_donorcluster %>% filter(cluster == 9) %>% select(-cluster)
a = gene_list(c_9,19.3)
cluster9_genes = Reduce(intersect, a)
# Cluster 10
c_10 = perc_express_donorcluster %>% filter(cluster == 10) %>% select(-cluster)
a = gene_list(c_10,6.5)
cluster10_genes = Reduce(intersect, a)
# Cluster 11
c_11 = perc_express_donorcluster %>% filter(cluster == 11) %>% select(-cluster)
a = gene_list(c_11,0)
cluster11_genes = Reduce(intersect, a) # only 400 shared genes between these
# Cluster 12
c_12 = perc_express_donorcluster %>% filter(cluster == 12) %>% select(-cluster)
a = gene_list(c_12,0)
cluster12_genes = Reduce(intersect, a) # only 275 shared genes between donors
rm(a)

# create one dataframe with all clusters genes, renaming to each cluster

df_list = list(c_0=c_0,c_1=c_1,c_2=c_2,c_3=c_3,c_4=c_4,c_5=c_5,c_6=c_6,c_7=c_7,c_8=c_8,c_9=c_9,c_10=c_10,c_11=c_11,c_12=c_12)
genes_list = list(c_0=cluster0_genes,c_1=cluster1_genes,c_2=cluster2_genes,c_3=cluster3_genes,c_4=cluster4_genes,c_5=cluster5_genes,c_6=cluster6_genes,
                  c_7=cluster7_genes,c_8=cluster8_genes,c_9=cluster9_genes,c_10=cluster10_genes,c_11=cluster11_genes,c_12=cluster12_genes)

for(df_name in names(df_list)) {
  df_list[[df_name]] = df_list[[df_name]] %>% select(donor, genes_list[[df_name]])
  df_list[[df_name]] = df_list[[df_name]] %>%
    rename_with(.fn = ~paste(., df_name, sep = "_"), .cols = c(genes_list[[df_name]]))
}

rm(c_0,c_1,c_2,c_3,c_4,c_5,c_6,c_7,c_8,c_9,c_10,c_11,c_12)

expressions_df = as.data.frame(df_list %>% reduce(left_join, by = "donor"))
```

```{r outcomes derive, include = FALSE}
# outcomes data
read_excel_allsheets <- function(filename, tibble = FALSE) {
    # I prefer straight data.frames
    # but if you like tidyverse tibbles (the default with read_excel)
    # then just pass tibble = TRUE
    sheets <- readxl::excel_sheets(filename)
    x <- lapply(sheets, function(X) readxl::read_excel(filename, sheet = X))
    if(!tibble) x <- lapply(x, as.data.frame)
    names(x) <- sheets
    x
}


a1ccpep = read_excel_allsheets("S:/Laura/BDC/Projects/Howard Davidson/R01/Data_Raw/Davidson pilot data/IDDA1C_Cpep.xlsx")
# remove all df as it looks incomplete
a1ccpep = a1ccpep[-1]
############# mixed part ################################################################################################################
# change to numeric
outcomes_df <- Reduce(function(x, y) merge(x, y, all = TRUE), a1ccpep)
outcomes_df = outcomes_df %>% filter(!is.na(`HWD ID`)) %>% mutate(a1c = as.numeric( case_when(A1c_Value == "NULL" ~ NA,
                                                                        TRUE ~ A1c_Value)),
                                                        bmi = as.numeric( case_when(BMI == "NULL" ~ NA,
                                                                        TRUE ~ BMI)))
# add in visit num and baseline values
outcomes_df = outcomes_df %>% arrange(`HWD ID`, as.Date(VisitDate)) %>% group_by(`HWD ID`) %>% mutate(visit = row_number(),
                                                                                                      est_cpep = case_when(`Est C-pep` >= 0 ~ `Est C-pep` ,
                                                                                                                        `Est C-pep` < 0 ~ 0 ),
                                                                                                      hwdid=`HWD ID`)

############################################################################################################################################
# change to numeric
a1c_df <- Reduce(function(x, y) merge(x, y, all = TRUE), a1ccpep)
a1c_df = a1c_df %>% filter(!is.na(`HWD ID`)) %>% mutate(a1c = as.numeric( case_when(A1c_Value == "NULL" ~ NA,
                                                                        TRUE ~ A1c_Value))) ; rm(a1ccpep)
# add in outcome vars
a1c_df = a1c_df %>% arrange(`HWD ID`, as.Date(VisitDate)) %>% group_by(`HWD ID`) %>% mutate(visit = row_number(),
                                                                                            a1c_g7 = ifelse(a1c > 7, "G7","L7")) # visit num
a1c_df = data.table(a1c_df, key = '`HWD ID`')
a1c_df = a1c_df[, days_from_v1 :=  cumsum(c(0, diff(as.Date(VisitDate)))), by=`HWD ID`]# days from baseline
a1c_df = a1c_df %>% mutate(y1_elig = ifelse(abs(days_from_v1- 365) < 90, 1, 0),
                           abs_y1_days = ifelse(abs(days_from_v1- 365) < 90, abs(days_from_v1- 365), NA))

a1c_y1 = a1c_df %>% group_by(`HWD ID`) %>% summarise(min_y1_days = min(abs_y1_days, na.rm = T))
a1c_y1 = a1c_df %>% group_by(`HWD ID`) %>% summarise(min_y1_days = min(abs_y1_days, na.rm = T))
a1c_y1$min_y1_days[is.infinite(a1c_y1$min_y1_days)] = NA
a1c_df = full_join(a1c_df, a1c_y1); rm(a1c_y1)
a1c_df = a1c_df %>% mutate(yr1a1c = ifelse(abs_y1_days == min_y1_days, a1c, NA))
a1c_df = a1c_df %>% group_by(`HWD ID`) %>% fill(yr1a1c)


avg_a1c = a1c_df %>% group_by(`HWD ID`) %>% summarise(meana1c = mean(a1c, na.rm = T),
                                                      firsta1c = first(a1c, na_rm = T),
                                                      maxa1c = max(a1c, na.rm=T),
                                                      yr1a1c = mean(yr1a1c, na.rm = T)) %>% mutate(id = `HWD ID`)
# add in slopes and auc
a1c_slopes = a1c_df %>% group_by(`HWD ID`) %>% do(tidy(lm(a1c ~ days_from_v1, data = .))) %>% filter(term == "days_from_v1") %>% select(`HWD ID`, estimate) %>% rename(a1c_slope = estimate)
avg_a1c = left_join(avg_a1c, a1c_slopes) ; rm(a1c_slopes)

avg_a1c$a1c_auc = sapply(split(a1c_df,a1c_df$`HWD ID`),function(df) trapz(df$days_from_v1,df$a1c))

# dichotomize
avg_a1c = avg_a1c %>% mutate(avg_a1c_g_med = ifelse(meana1c > median(meana1c,na.rm = T), "G_med", " L_med"),
                             first_a1c_g_med = ifelse(firsta1c > median(firsta1c,na.rm = T), "G_med", " L_med"),
                             y1_a1c_g_med = ifelse(yr1a1c > median(yr1a1c,na.rm = T), "G_med", " L_med"),
                             max_a1c_g_med = ifelse(maxa1c > median(maxa1c,na.rm = T), "G_med", " L_med"),
                             slope_a1c_g_med = ifelse(a1c_slope > median(a1c_slope,na.rm = T), "G_med", " L_med"),
                             auc_a1c_g_med = ifelse(a1c_auc > median(a1c_auc,na.rm = T), "G_med", " L_med"))


df2@meta.data = left_join(df2@meta.data, avg_a1c,  by=c('donor_number'='id'))
# View(df2@meta.data)

# add in other outcomes
# IDDA1c
avg_idda1c = a1c_df %>% group_by(`HWD ID`) %>% summarise(mean_idda1c = mean(IDDA1C, na.rm = T)) %>% mutate(id = `HWD ID`)
# add in slopes and auc
idda1c_slopes = a1c_df %>% group_by(`HWD ID`) %>% do(tidy(lm(IDDA1C ~ days_from_v1, data = .))) %>% filter(term == "days_from_v1") %>% 
  select(`HWD ID`, estimate) %>% rename(idda1c_slope = estimate)
avg_idda1c = left_join(avg_idda1c, idda1c_slopes) ; rm(idda1c_slopes)

avg_idda1c$idda1c_auc = sapply(split(a1c_df,a1c_df$`HWD ID`),function(df) trapz(df$days_from_v1,df$IDDA1C))

# dichotomize
avg_idda1c = avg_idda1c %>% mutate(avg_idda1c_g_med = ifelse(mean_idda1c > median(mean_idda1c,na.rm = T), "G_med", " L_med"),
                             slope_idda1c_g_med = ifelse(idda1c_slope > median(idda1c_slope,na.rm = T), "G_med", " L_med"),
                             auc_idda1c_g_med = ifelse(idda1c_auc > median(idda1c_auc,na.rm = T), "G_med", " L_med"))

df2@meta.data = left_join(df2@meta.data, avg_idda1c,  by=c('donor_number'='id'))


# total daily insulin dose
a1c_df$dose_udk = a1c_df$`U/day/kg`
avg_dose_udk = a1c_df %>% group_by(`HWD ID`) %>% summarise(mean_dose_udk = mean(dose_udk, na.rm = T)) %>% mutate(id = `HWD ID`)
# add in slopes and auc
dose_udk_slopes = a1c_df %>% group_by(`HWD ID`) %>% do(tidy(lm(dose_udk ~ days_from_v1, data = .))) %>% filter(term == "days_from_v1") %>% 
  select(`HWD ID`, estimate) %>% rename(dose_udk_slope = estimate)
avg_dose_udk = left_join(avg_dose_udk, dose_udk_slopes) ; rm(dose_udk_slopes)

avg_dose_udk$dose_udk_auc = sapply(split(a1c_df,a1c_df$`HWD ID`),function(df) trapz(df$days_from_v1,df$dose_udk))

# dichotomize
avg_dose_udk = avg_dose_udk %>% mutate(avg_dose_udk_g_med = ifelse(mean_dose_udk > median(mean_dose_udk,na.rm = T), "G_med", " L_med"),
                             slope_dose_udk_g_med = ifelse(dose_udk_slope > median(dose_udk_slope,na.rm = T), "G_med", " L_med"),
                             auc_dose_udk_g_med = ifelse(dose_udk_auc > median(dose_udk_auc,na.rm = T), "G_med", " L_med"))

df2@meta.data = left_join(df2@meta.data, avg_dose_udk,  by=c('donor_number'='id'))

# cpep
a1c_df = a1c_df %>% mutate(est_cpep = ifelse(`Est C-pep` > 0, `Est C-pep`, 0))
avg_est_cpep = a1c_df %>% group_by(`HWD ID`) %>% summarise(mean_est_cpep = mean(est_cpep, na.rm = T)) %>% mutate(id = `HWD ID`)
# add in slopes and auc
est_cpep_slopes = a1c_df %>% group_by(`HWD ID`) %>% do(tidy(lm(est_cpep ~ days_from_v1, data = .))) %>% filter(term == "days_from_v1") %>% 
  select(`HWD ID`, estimate) %>% rename(est_cpep_slope = estimate)
avg_est_cpep = left_join(avg_est_cpep, est_cpep_slopes) ; rm(est_cpep_slopes)

avg_est_cpep$est_cpep_auc = sapply(split(a1c_df,a1c_df$`HWD ID`),function(df) trapz(df$days_from_v1,df$est_cpep))

# dichotomize
avg_est_cpep = avg_est_cpep %>% mutate(avg_est_cpep_g_med = ifelse(mean_est_cpep > median(mean_est_cpep,na.rm = T), "G_med", " L_med"),
                             slope_est_cpep_g_med = ifelse(est_cpep_slope > median(est_cpep_slope,na.rm = T), "G_med", " L_med"),
                             auc_est_cpep_g_med = ifelse(est_cpep_auc > median(est_cpep_auc,na.rm = T), "G_med", " L_med"))



########## MIXED MODELS COMPONENT #################################################################


# IDDA1c
idda1c_df_mixed = outcomes_df %>% select(`HWD ID`, IDDA1C, VisitDate, visit, bmi, Gender, `Age at visit`)
idda1c_df_mixed = idda1c_df_mixed %>% group_by(`HWD ID`) %>% mutate(first_visit = case_when(visit== 1 ~ VisitDate),
                                                            idda1c_bl = case_when(visit==1~IDDA1C)) %>% fill(first_visit,idda1c_bl) %>% ungroup() 
# group into 6month bins
idda1c_df_mixed = idda1c_df_mixed %>% mutate(month_from_bl = as.duration(first_visit %--% VisitDate)/dmonths(1))
idda1c_df_mixed = idda1c_df_mixed %>% mutate(visit_binned = floor(month_from_bl/6))

analysis_idda1c_mixed = idda1c_df_mixed %>% group_by(`HWD ID`, visit_binned) %>% mutate(IDDA1C_binned = mean(IDDA1C, na.rm=T),
                                                                 bmi_binned = mean(bmi, na.rm=T),
                                                                 age_binned = mean(`Age at visit`, na.rm=T)) %>%
  select(`HWD ID`, IDDA1C_binned, bmi_binned, age_binned, Gender, idda1c_bl, visit_binned) %>% unique()

##################################################################################################

# est cpep
est_cpep_df_mixed = outcomes_df %>% select(`HWD ID`, est_cpep, VisitDate, visit, bmi, Gender, `Age at visit`)
est_cpep_df_mixed = est_cpep_df_mixed %>% group_by(`HWD ID`) %>% mutate(first_visit = case_when(visit== 1 ~ VisitDate),
                                                            est_cpep_bl = case_when(visit==1~est_cpep)) %>% fill(first_visit,est_cpep_bl) %>% ungroup() 
# group into 6month bins
est_cpep_df_mixed = est_cpep_df_mixed %>% mutate(month_from_bl = as.duration(first_visit %--% VisitDate)/dmonths(1))
est_cpep_df_mixed = est_cpep_df_mixed %>% mutate(visit_binned = floor(month_from_bl/6))

analysis_est_cpep_mixed = est_cpep_df_mixed %>% group_by(`HWD ID`, visit_binned) %>% mutate(est_cpep_binned = mean(est_cpep, na.rm=T),
                                                                 bmi_binned = mean(bmi, na.rm=T),
                                                                 age_binned = mean(`Age at visit`, na.rm=T)) %>%
  select(`HWD ID`, est_cpep_binned, bmi_binned, age_binned, Gender, est_cpep_bl, visit_binned) %>% unique()


##################################################################################################
outcomes_df$dose_udk = outcomes_df$`U/day/kg`
# daily dose
dose_udk_df_mixed = outcomes_df %>% select(`HWD ID`, dose_udk, VisitDate, visit, bmi, Gender, `Age at visit`)
dose_udk_df_mixed = dose_udk_df_mixed %>% group_by(`HWD ID`) %>% mutate(first_visit = case_when(visit== 1 ~ VisitDate),
                                                            dose_udk_bl = case_when(visit==1~dose_udk)) %>% fill(first_visit,dose_udk_bl) %>% ungroup() 
# group into 6month bins
dose_udk_df_mixed = dose_udk_df_mixed %>% mutate(month_from_bl = as.duration(first_visit %--% VisitDate)/dmonths(1))
dose_udk_df_mixed = dose_udk_df_mixed %>% mutate(visit_binned = floor(month_from_bl/6))

analysis_dose_udk_mixed = dose_udk_df_mixed %>% group_by(`HWD ID`, visit_binned) %>% mutate(dose_udk_binned = mean(dose_udk, na.rm=T),
                                                                 bmi_binned = mean(bmi, na.rm=T),
                                                                 age_binned = mean(`Age at visit`, na.rm=T)) %>%
  select(`HWD ID`, dose_udk_binned, bmi_binned, age_binned, Gender, dose_udk_bl, visit_binned) %>% unique()


##################################################################################################

# a1c
a1c_df_mixed = outcomes_df %>% select(`HWD ID`, a1c, VisitDate, visit, bmi, Gender, `Age at visit`)
a1c_df_mixed = a1c_df_mixed %>% group_by(`HWD ID`) %>% mutate(first_visit = case_when(visit== 1 ~ VisitDate),
                                                            a1c_bl = case_when(visit==1~a1c)) %>% fill(first_visit,a1c_bl) %>% ungroup() 
# group into 6month bins
a1c_df_mixed = a1c_df_mixed %>% mutate(month_from_bl = as.duration(first_visit %--% VisitDate)/dmonths(1))
a1c_df_mixed = a1c_df_mixed %>% mutate(visit_binned = floor(month_from_bl/6))

analysis_a1c_mixed = a1c_df_mixed %>% group_by(`HWD ID`, visit_binned) %>% mutate(a1c_binned = mean(a1c, na.rm=T),
                                                                 bmi_binned = mean(bmi, na.rm=T),
                                                                 age_binned = mean(`Age at visit`, na.rm=T)) %>%
  select(`HWD ID`, a1c_binned, bmi_binned, age_binned, Gender, a1c_bl, visit_binned) %>% unique()



##################################################################################################
```

```{r variable selection and elnet, include = FALSE}


# donor and hwd id match
ids = df2@meta.data %>% select(donor, `HWD ID.y`) %>% unique()
colnames(ids) = c("donor", "HWD ID")
avg_idda1c = full_join(avg_idda1c, ids)

# iddA1c Slope ##################################################################################################


# merge with gene expressions
expressions_df$donor = factor(expressions_df$donor)
analysis_idda1c = full_join(avg_idda1c, expressions_df)
analysis_idda1c_slope = analysis_idda1c %>%  filter(!is.na(donor)) %>% select(idda1c_slope, colnames(expressions_df))


#pare the list of 6000 genes down with univariate regression, p <.1
models_simple = list()
# with apply:
coefs_mat = expand.grid(c("idda1c_slope"), colnames(expressions_df)[-1])

mods = apply(coefs_mat, 1, function(row) {
  lm(as.formula(paste(row[1], "~", row[2])), data = analysis_idda1c_slope)
})
names(mods) = with(coefs_mat, paste(Var1, "vs", Var2))
coefs = lapply(mods, tidy, simplify = F)


# combine
uni_res = dplyr::bind_rows(coefs, .id = "mod")
uni_res = uni_res %>% filter(p.value < 0.1 & term != "(Intercept)")
new_genes = uni_res$term
#new_genes =  make.names(new_genes , unique = T, allow_ = F)


# scale and center the data
set.seed(1017)
analysis_idda1c_slope = analysis_idda1c_slope %>% select(-donor)
recipe_idda1c_slope = recipe(analysis_idda1c_slope) %>% update_role(everything()) %>% update_role(idda1c_slope, new_role = "outcome")
recipe_idda1c_slope = recipe_idda1c_slope %>%
  step_zv(all_predictors()) %>%
  step_nzv(all_predictors()) %>%
  step_YeoJohnson(all_predictors()) %>%
  step_normalize(all_predictors()) %>%
  prep()
transformed_analysis_idda1c_slope <- bake(recipe_idda1c_slope,analysis_idda1c_slope)
predictors_idda1c_slope_en_large = colnames(transformed_analysis_idda1c_slope)[-1]
predictors_idda1c_slope_en = intersect(predictors_idda1c_slope_en_large, new_genes)


idda1c_slope_en_small = easy_elasticnet(data = transformed_analysis_idda1c_slope, outcome = "idda1c_slope",cv_method = "loo",
                              predictors = predictors_idda1c_slope_en, model_type = "gaussian", out="1se.error")

##################################################################################################
avg_est_cpep = full_join(avg_est_cpep, ids)
analysis_est_cpep = full_join(avg_est_cpep, expressions_df)
analysis_est_cpep_slope = analysis_est_cpep %>%  filter(!is.na(donor)) %>% select(est_cpep_slope, colnames(expressions_df))


#pare the list of 6000 genes down with univariate regression, p <.1
models_simple = list()
# with apply:
coefs_mat = expand.grid(c("est_cpep_slope"), colnames(expressions_df)[-1])

mods = apply(coefs_mat, 1, function(row) {
  lm(as.formula(paste(row[1], "~", row[2])), data = analysis_est_cpep_slope)
})
names(mods) = with(coefs_mat, paste(Var1, "vs", Var2))
coefs = lapply(mods, tidy, simplify = F)


# combine
uni_res = dplyr::bind_rows(coefs, .id = "mod")
uni_res = uni_res %>% filter(p.value < 0.1 & term != "(Intercept)")
new_genes = uni_res$term
#new_genes =  make.names(new_genes , unique = T, allow_ = F)


# scale and center the data
set.seed(1017)
analysis_est_cpep_slope = analysis_est_cpep_slope %>% select(-donor)
recipe_est_cpep_slope = recipe(analysis_est_cpep_slope) %>% update_role(everything()) %>% update_role(est_cpep_slope, new_role = "outcome")
recipe_est_cpep_slope = recipe_est_cpep_slope %>%
  step_zv(all_predictors()) %>%
  step_nzv(all_predictors()) %>%
  step_YeoJohnson(all_predictors()) %>%
  step_normalize(all_predictors()) %>%
  prep()
transformed_analysis_est_cpep_slope <- bake(recipe_est_cpep_slope,analysis_est_cpep_slope)
predictors_est_cpep_slope_en_large = colnames(transformed_analysis_est_cpep_slope)[-1]
predictors_est_cpep_slope_en = intersect(predictors_est_cpep_slope_en_large, new_genes)


est_cpep_slope_en_small = easy_elasticnet(data = transformed_analysis_est_cpep_slope, outcome = "est_cpep_slope",cv_method = "loo",
                              predictors = predictors_est_cpep_slope_en, model_type = "gaussian", out="1se.error")

##################################################################################################
avg_dose_udk = full_join(avg_dose_udk, ids)

analysis_dose_udk = full_join(avg_dose_udk, expressions_df)
analysis_dose_udk_slope = analysis_dose_udk %>%  filter(!is.na(donor)) %>% select(dose_udk_slope, colnames(expressions_df))


#pare the list of 6000 genes down with univariate regression, p <.1
models_simple = list()
# with apply:
coefs_mat = expand.grid(c("dose_udk_slope"), colnames(expressions_df)[-1])

mods = apply(coefs_mat, 1, function(row) {
  lm(as.formula(paste(row[1], "~", row[2])), data = analysis_dose_udk_slope)
})
names(mods) = with(coefs_mat, paste(Var1, "vs", Var2))
coefs = lapply(mods, tidy, simplify = F)


# combine
uni_res = dplyr::bind_rows(coefs, .id = "mod")
uni_res = uni_res %>% filter(p.value < 0.1 & term != "(Intercept)")
new_genes = uni_res$term
#new_genes =  make.names(new_genes , unique = T, allow_ = F)


# scale and center the data
set.seed(1017)
analysis_dose_udk_slope = analysis_dose_udk_slope %>% select(-donor)
recipe_dose_udk_slope = recipe(analysis_dose_udk_slope) %>% update_role(everything()) %>% update_role(dose_udk_slope, new_role = "outcome")
recipe_dose_udk_slope = recipe_dose_udk_slope %>%
  step_zv(all_predictors()) %>%
  step_nzv(all_predictors()) %>%
  step_YeoJohnson(all_predictors()) %>%
  step_normalize(all_predictors()) %>%
  prep()
transformed_analysis_dose_udk_slope <- bake(recipe_dose_udk_slope,analysis_dose_udk_slope)
predictors_dose_udk_slope_en_large = colnames(transformed_analysis_dose_udk_slope)[-1]
predictors_dose_udk_slope_en = intersect(predictors_dose_udk_slope_en_large, new_genes)


dose_udk_slope_en_small = easy_elasticnet(data = transformed_analysis_dose_udk_slope, outcome = "dose_udk_slope",cv_method = "loo",
                              predictors = predictors_dose_udk_slope_en, model_type = "gaussian", out="1se.error")
##################################################################################################
avg_a1c = full_join(avg_a1c, ids)

analysis_a1c = full_join(avg_a1c, expressions_df)
analysis_a1c_slope = analysis_a1c %>%  filter(!is.na(donor)) %>% select(a1c_slope, colnames(expressions_df))


#pare the list of 6000 genes down with univariate regression, p <.1
models_simple = list()
# with apply:
coefs_mat = expand.grid(c("a1c_slope"), colnames(expressions_df)[-1])

mods = apply(coefs_mat, 1, function(row) {
  lm(as.formula(paste(row[1], "~", row[2])), data = analysis_a1c_slope)
})
names(mods) = with(coefs_mat, paste(Var1, "vs", Var2))
coefs = lapply(mods, tidy, simplify = F)


# combine
uni_res = dplyr::bind_rows(coefs, .id = "mod")
uni_res = uni_res %>% filter(p.value < 0.1 & term != "(Intercept)")
new_genes = uni_res$term
#new_genes =  make.names(new_genes , unique = T, allow_ = F)


# scale and center the data
set.seed(1017)
analysis_a1c_slope = analysis_a1c_slope %>% select(-donor)
recipe_a1c_slope = recipe(analysis_a1c_slope) %>% update_role(everything()) %>% update_role(a1c_slope, new_role = "outcome")
recipe_a1c_slope = recipe_a1c_slope %>%
  step_zv(all_predictors()) %>%
  step_nzv(all_predictors()) %>%
  step_YeoJohnson(all_predictors()) %>%
  step_normalize(all_predictors()) %>%
  prep()
transformed_analysis_a1c_slope <- bake(recipe_a1c_slope,analysis_a1c_slope)
predictors_a1c_slope_en_large = colnames(transformed_analysis_a1c_slope)[-1]
predictors_a1c_slope_en = intersect(predictors_a1c_slope_en_large, new_genes)


a1c_slope_en_small = easy_elasticnet(data = transformed_analysis_a1c_slope, outcome = "a1c_slope",cv_method = "loo",
                              predictors = predictors_a1c_slope_en, model_type = "gaussian", out="1se.error")

##################################################################################################

```


```{r mixed model, include = F}
library(sjPlot)
#change to factor
expressions_df$donor = as.factor(expressions_df$donor)
##################################################################################################
# idda1c mixed df
analysis_idda1c_mixed = left_join(analysis_idda1c_mixed, ids)%>% ungroup() %>% select(-`HWD ID`)
expressions_df_idda1c_mixed = left_join(analysis_idda1c_mixed, expressions_df)
#table(expressions_df_idda1c$visit_binned)

expressions_df_idda1c_mixed = expressions_df_idda1c_mixed %>% filter(visit_binned < 5)
# select vars chosen in tims elastic net
expressions_df_idda1c_mixed = expressions_df_idda1c_mixed %>% dplyr::select(IDDA1C_binned,idda1c_bl,donor,visit_binned,bmi_binned,Gender,age_binned, all_of(idda1c_slope_en_small))
expressions_df_idda1c_mixed = expressions_df_idda1c_mixed %>% filter(!is.na(donor))

# create form
idda1c_coefs_mm = colnames(expressions_df_idda1c_mixed %>% dplyr::select(-donor)) 
idda1c_mm_form = as.formula(v2frm(idda1c_coefs_mm, y = 1L))



idda1c_mixed_mod = lmer(IDDA1C_binned ~ idda1c_bl + visit_binned +Gender + bmi_binned +age_binned + RPL35A_c_5 + RPS3_c_7 + RPS6_c_8 + HNRNPDL_c_10 + RPS10_c_12 + HSP90AA1_c_12 + RPL3_c_12 + (1|donor),
            data = expressions_df_idda1c_mixed)
# summary(idda1c_mixed_mod)

# plot predicted values of idda1c
idda1c_plot_df = expressions_df_idda1c_mixed%>% na.omit() %>% dplyr::mutate(pred_idda1c = predict(idda1c_mixed_mod, newdata=.,re.form=~(1|donor)))

##################################################################################################
# est_cpep mixed df
analysis_est_cpep_mixed = left_join(analysis_est_cpep_mixed, ids)%>% ungroup() %>% select(-`HWD ID`)
expressions_df_est_cpep_mixed = left_join(analysis_est_cpep_mixed, expressions_df)
#table(expressions_df_est_cpep$visit_binned)

expressions_df_est_cpep_mixed = expressions_df_est_cpep_mixed %>% filter(visit_binned < 5)
# select vars chosen in tims elastic net
expressions_df_est_cpep_mixed = expressions_df_est_cpep_mixed %>% dplyr::select(est_cpep_binned,est_cpep_bl,donor,visit_binned,bmi_binned,Gender,age_binned, all_of(est_cpep_slope_en_small))
expressions_df_est_cpep_mixed = expressions_df_est_cpep_mixed %>% filter(!is.na(donor))

# create form
est_cpep_coefs_mm = colnames(expressions_df_est_cpep_mixed %>% dplyr::select(-donor)) 
est_cpep_mm_form = as.formula(v2frm(est_cpep_coefs_mm, y = 1L))



est_cpep_mixed_mod = lmer(est_cpep_binned ~ est_cpep_bl + visit_binned +Gender + bmi_binned +age_binned + RPL35A_c_5  + HNRNPDL_c_10 +RPS10_c_12  + (1|donor),
            data = expressions_df_est_cpep_mixed)
# summary(est_cpep_mixed_mod)

# plot predicted values of est_cpep
est_cpep_plot_df = expressions_df_est_cpep_mixed%>% na.omit() %>% dplyr::mutate(pred_est_cpep = predict(est_cpep_mixed_mod, newdata=.,re.form=~(1|donor)))

##################################################################################################
# dose_udk mixed df
analysis_dose_udk_mixed = left_join(analysis_dose_udk_mixed, ids)%>% ungroup() %>% select(-`HWD ID`)
expressions_df_dose_udk_mixed = left_join(analysis_dose_udk_mixed, expressions_df)
#table(expressions_df_dose_udk$visit_binned)

expressions_df_dose_udk_mixed = expressions_df_dose_udk_mixed %>% filter(visit_binned < 5)
# select vars chosen in tims elastic net
expressions_df_dose_udk_mixed = expressions_df_dose_udk_mixed %>% dplyr::select(dose_udk_binned,dose_udk_bl,donor,visit_binned,bmi_binned,Gender,age_binned, all_of(dose_udk_slope_en_small))
expressions_df_dose_udk_mixed = expressions_df_dose_udk_mixed %>% filter(!is.na(donor))

# create form
dose_udk_coefs_mm = colnames(expressions_df_dose_udk_mixed %>% dplyr::select(-donor)) 
dose_udk_mm_form = as.formula(v2frm(dose_udk_coefs_mm, y = 1L))



dose_udk_mixed_mod = lmer(dose_udk_binned ~ dose_udk_bl + visit_binned +Gender + bmi_binned +age_binned +
                            RPS3_c_7+ RPS6_c_8 + RPS10_c_12 + HSP90AA1_c_12+ RPL3_c_12  + (1|donor),
            data = expressions_df_dose_udk_mixed)
# summary(dose_udk_mixed_mod)

# plot predicted values of dose_udk
dose_udk_plot_df = expressions_df_dose_udk_mixed%>% na.omit() %>% dplyr::mutate(pred_dose_udk = predict(dose_udk_mixed_mod, newdata=.,re.form=~(1|donor)))

##################################################################################################
# a1c mixed df
analysis_a1c_mixed = left_join(analysis_a1c_mixed, ids)%>% ungroup() %>% select(-`HWD ID`)
expressions_df_a1c_mixed = left_join(analysis_a1c_mixed, expressions_df)
#table(expressions_df_a1c$visit_binned)

expressions_df_a1c_mixed = expressions_df_a1c_mixed %>% filter(visit_binned < 5)
# select vars chosen in tims elastic net
expressions_df_a1c_mixed = expressions_df_a1c_mixed %>% dplyr::select(a1c_binned,a1c_bl,donor,visit_binned,bmi_binned,Gender,age_binned, all_of(a1c_slope_en_small))
expressions_df_a1c_mixed = expressions_df_a1c_mixed %>% filter(!is.na(donor))

# create form
a1c_coefs_mm = colnames(expressions_df_a1c_mixed %>% dplyr::select(-donor)) 
a1c_mm_form = as.formula(v2frm(a1c_coefs_mm, y = 1L))

a1c_geneform = paste0(a1c_slope_en_small, collapse = "+")
a1cform = as.formula(paste0("a1c_binned ~ a1c_bl + visit_binned +Gender + bmi_binned +age_binned +",a1c_geneform," + (1|donor)"))
a1c_mixed_mod = lmer(a1cform,
            data = expressions_df_a1c_mixed)
# summary(a1c_mixed_mod)

# plot predicted values of a1c
a1c_plot_df = expressions_df_a1c_mixed%>% na.omit() %>% dplyr::mutate(pred_a1c = predict(a1c_mixed_mod, newdata=.,re.form=~(1|donor)))

##################################################################################################

```

# Methods

Due to the limited sample size, the LASSO models using repeated measures (i.e., not summary measures like slope) would not converge. As a result, we changed the modeling strategy:

1)	Percent expressions were calculated for each cluster by donor; about 500 genes from each cluster with the highest percent expression among all donors were used for initial variable selection.
2)	We performed initial univariate screening.
a.	Outcomes for variable selection were defined as the linear rate of change over time for IDDA1c (Est Daily Dose, A1c, Est Cpep also derived).
b.	Univariate regressions were performed on the outcomes for each gene in the initial variable selection list. Genes associated with the outcomes at the 0.1 significance level were included in the elasticnet variable selection.
3)	The elastic net algorithm was applied to the genes identified in univariate screening, using the slope of the outcome measures as the dependent variable (i.e., not mixed-effect repeated measures models).
a.	Leave One Out cross validation was performed for elasticnet variable selection methods. Models within 1 SE were examined, with the best model being chosen by MSE.
4)	The genes selected by the elastic net algorithm were included in a mixed model with random intercept for subject, adjusting for gender, age, bmi, time and baseline outcome values. We needed to “bin” the visits into 6-month intervals rather than 3-month intervals due to sparse data. Any repeat measures within a bin were averaged. 
a.	Model summaries were produced, and model-predicted values were compared to observed values in the figures below.



# Analysis

## IDDA1c Analysis

### Model summary

```{r}
summary(idda1c_mixed_mod)
```

### Predicted vs Actual Plots

Red indicates predicted IDDA1C via the mixed model and Black indicates actual IDDA1C. 

```{r}
plot_model(idda1c_mixed_mod, terms = "visit_binned",type = "pred", color = "black") + theme_classic() +
   geom_smooth(aes(y = IDDA1C_binned, x = visit_binned), 
                                data = idda1c_plot_df, method = "lm", color = "red") + 
  xlab("Time From First Visit (6 Month Increments)") + ylab("6 Month Average IDDA1C") + labs(title = "")
```

Reference line of slope 1 included.

```{r}
idda1c_plot_df %>% ggplot(aes(x = IDDA1C_binned, y = pred_idda1c)) +
  geom_abline(intercept = 0, slope = 1) + 
  geom_point() + theme_classic()+ 
  xlab("IDDA1C Actual") + ylab("IDDA1C Predicted") + labs(title = "")
```

Predicted IDDA1C over actual IDDA1C distributions
```{r}
idda1c_plot_df%>%
  ggplot(aes(x = visit_binned, y = IDDA1C_binned)) +
  geom_point()+
  geom_smooth(aes(y = pred_idda1c), method = "lm",color = "red") + theme_classic()
```

## Est CPEP Analysis

### Model summary

```{r}
summary(est_cpep_mixed_mod)
```

### Predicted vs Actual Plots

Red indicates predicted est_cpep via the mixed model and Black indicates actual est_cpep. 

```{r}
plot_model(est_cpep_mixed_mod, terms = "visit_binned",type = "pred", color = "black") + theme_classic() +
   geom_smooth(aes(y = est_cpep_binned, x = visit_binned), 
                                data = est_cpep_plot_df, method = "lm", color = "red") + 
  xlab("Time From First Visit (6 Month Increments)") + ylab("6 Month Average est_cpep") + labs(title = "")
```

Reference line of slope 1 included.

```{r}
est_cpep_plot_df %>% ggplot(aes(x = est_cpep_binned, y = pred_est_cpep)) +
  geom_abline(intercept = 0, slope = 1) + 
  geom_point() + theme_classic()+ 
  xlab("est_cpep Actual") + ylab("est_cpep Predicted") + labs(title = "")
```

Predicted est_cpep over actual est_cpep distributions
```{r}
est_cpep_plot_df%>%
  ggplot(aes(x = visit_binned, y = est_cpep_binned)) +
  geom_point()+
  geom_smooth(aes(y = pred_est_cpep), method = "lm",color = "red") + theme_classic()
```

## dose_udk Analysis

### Model summary

```{r}
summary(dose_udk_mixed_mod)
```

### Predicted vs Actual Plots

Red indicates predicted dose_udk via the mixed model and Black indicates actual dose_udk. 

```{r}
plot_model(dose_udk_mixed_mod, terms = "visit_binned",type = "pred", color = "black") + theme_classic() +
   geom_smooth(aes(y = dose_udk_binned, x = visit_binned), 
                                data = dose_udk_plot_df, method = "lm", color = "red") + 
  xlab("Time From First Visit (6 Month Increments)") + ylab("6 Month Average dose_udk") + labs(title = "")
```

Reference line of slope 1 included.

```{r}
dose_udk_plot_df %>% ggplot(aes(x = dose_udk_binned, y = pred_dose_udk)) +
  geom_abline(intercept = 0, slope = 1) + 
  geom_point() + theme_classic()+ 
  xlab("dose_udk Actual") + ylab("dose_udk Predicted") + labs(title = "")
```

Predicted dose_udk over actual dose_udk distributions
```{r}
dose_udk_plot_df%>%
  ggplot(aes(x = visit_binned, y = dose_udk_binned)) +
  geom_point()+
  geom_smooth(aes(y = pred_dose_udk), method = "lm",color = "red") + theme_classic()
```

## a1c Analysis

### Model summary

```{r}
summary(a1c_mixed_mod)
```

### Predicted vs Actual Plots

Red indicates predicted a1c via the mixed model and Black indicates actual a1c. 

```{r}
plot_model(a1c_mixed_mod, terms = "visit_binned",type = "pred", color = "black") + theme_classic() +
   geom_smooth(aes(y = a1c_binned, x = visit_binned), 
                                data = a1c_plot_df, method = "lm", color = "red") + 
  xlab("Time From First Visit (6 Month Increments)") + ylab("6 Month Average a1c") + labs(title = "")
```

Reference line of slope 1 included.

```{r}
a1c_plot_df %>% ggplot(aes(x = a1c_binned, y = pred_a1c)) +
  geom_abline(intercept = 0, slope = 1) + 
  geom_point() + theme_classic()+ 
  xlab("a1c Actual") + ylab("a1c Predicted") + labs(title = "")
```

Predicted a1c over actual a1c distributions
```{r}
a1c_plot_df%>%
  ggplot(aes(x = visit_binned, y = a1c_binned)) +
  geom_point()+
  geom_smooth(aes(y = pred_a1c), method = "lm",color = "red") + theme_classic()
```


# Appendix

- Percent expression cutoffs for each cluster (to pare gene list down to ~ 500):
c0: 22
c1: 28
c2: 26
c3: 28
c4: 32
c5: 29
c6: 21
c7: 26
c8: 26
c9: 19
c10: 6
c11: 0 (400 shared genes in cluster)
c12: 0 (275 shared genes in cluster)

- idda1c selected genes:  `r idda1c_slope_en_small`
- est cpep selected genes:  `r est_cpep_slope_en_small`
- dose udk selected genes:  `r dose_udk_slope_en_small`
- a1c selected genes:  `r a1c_slope_en_small`