---
title: ""
author: "Tim Vigers & Laura Pyle"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_depth: 5
    toc_float: yes
---

```{r setup, include=FALSE}
library(tidyverse)
library(Seurat) # need these for Differential expression testing
library(SeuratObject)
library(data.table)
library(broom)
library(pROC)
library(caTools)

library(caret)
library(tidymodels)
library(doParallel)
library(DataExplorer)
library(knitr)
library(ranger)
# update the dfs as change (changes in first chunk of version 3 on the Lambda)
#perc_express_cluster <- read.csv("S:/Laura/BDC/Projects/Howard Davidson/perc_express_cluster_TEST.csv")

# check donor cluster level
perc_express_donorcluster  <- read.csv("S:/Laura/BDC/Projects/Howard Davidson/perc_express_donorcluster_TEST.csv")
#rownames(df2) = colnames(perc_express_donorcluster)

# data
home_dir = "S:/Laura/BDC/Projects/Howard Davidson/R01/Data_Raw/Davidson pilot data/"
knitr::opts_knit$set(root.dir = home_dir)
#setwd("S:/Laura/BDC/Projects/Howard Davidson/R01/Data_Raw/Davidson pilot data/")
df2 = readRDS(paste0(home_dir, "OneDrive_1_6-28-2023/normalized.rds"))

```

```{r fun, include = FALSE}
gene_list = function(df, cutoff){
  # df is one cluster w/ all donors at perc_expr level
  dft = as.data.frame(t(df)[-1,])
  donors  = colnames(dft)
  donorgenes = list()
  for(i in donors){
    nam = dft %>% select(i)
    donor_vec = nam %>% filter(.[[1]] > cutoff)  %>% rownames()
    donorgenes[[i]] = donor_vec
  }
  return(donorgenes)
}

# Tims elnet
#source("D:/Repositories/shared-resources/Machine Learning/Tim - ElasticNet CV/easy_elasticnet.R")
source("C:/Repositories/shared-resources/Machine Learning/Tim - ElasticNet CV/easy_elasticnet_cs.R")
```

```{r col clusters, include = FALSE}
# 22.3 for 505 
# Cluster 0
c_0 = perc_express_donorcluster %>% filter(cluster == 0) %>% select(-cluster)
a = gene_list(c_0, 22.3) 
cluster0_genes = Reduce(intersect, a) 


# Cluster 1
c_1 = perc_express_donorcluster %>% filter(cluster == 1) %>% select(-cluster)
a = gene_list(c_1, 28) # 28 for 502
cluster1_genes = Reduce(intersect, a)

# Cluster 2
c_2 = perc_express_donorcluster %>% filter(cluster == 2) %>% select(-cluster)
a = gene_list(c_2, 26)
cluster2_genes = Reduce(intersect, a) #26% for 500 exactly hot dog

# Cluster 3
c_3 = perc_express_donorcluster %>% filter(cluster == 3) %>% select(-cluster)
a = gene_list(c_3, 28)
cluster3_genes = Reduce(intersect, a)
# Cluster 4
c_4 = perc_express_donorcluster %>% filter(cluster == 4) %>% select(-cluster)
a = gene_list(c_4, 31.6)
cluster4_genes = Reduce(intersect, a)
# Cluster 5
c_5 = perc_express_donorcluster %>% filter(cluster == 5) %>% select(-cluster)
a = gene_list(c_5, 29.4)
cluster5_genes = Reduce(intersect, a)
# Cluster 6
c_6 = perc_express_donorcluster %>% filter(cluster == 6) %>% select(-cluster)
a = gene_list(c_6, 21)
cluster6_genes = Reduce(intersect, a)
# Cluster 7
c_7 = perc_express_donorcluster %>% filter(cluster == 7) %>% select(-cluster)
a = gene_list(c_7,26)
cluster7_genes = Reduce(intersect, a)
# Cluster 8
c_8 = perc_express_donorcluster %>% filter(cluster == 8) %>% select(-cluster)
a = gene_list(c_8, 26.3)
cluster8_genes = Reduce(intersect, a)
# Cluster 9
c_9 = perc_express_donorcluster %>% filter(cluster == 9) %>% select(-cluster)
a = gene_list(c_9,19.3)
cluster9_genes = Reduce(intersect, a)
# Cluster 10
c_10 = perc_express_donorcluster %>% filter(cluster == 10) %>% select(-cluster)
a = gene_list(c_10,6.5)
cluster10_genes = Reduce(intersect, a)
# Cluster 11
c_11 = perc_express_donorcluster %>% filter(cluster == 11) %>% select(-cluster)
a = gene_list(c_11,0)
cluster11_genes = Reduce(intersect, a) # only 400 shared genes between these
# Cluster 12
c_12 = perc_express_donorcluster %>% filter(cluster == 12) %>% select(-cluster)
a = gene_list(c_12,0)
cluster12_genes = Reduce(intersect, a) # only 275 shared genes between donors
rm(a)

# create one dataframe with all clusters genes, renaming to each cluster

df_list = list(c_0=c_0,c_1=c_1,c_2=c_2,c_3=c_3,c_4=c_4,c_5=c_5,c_6=c_6,c_7=c_7,c_8=c_8,c_9=c_9,c_10=c_10,c_11=c_11,c_12=c_12)
genes_list = list(c_0=cluster0_genes,c_1=cluster1_genes,c_2=cluster2_genes,c_3=cluster3_genes,c_4=cluster4_genes,c_5=cluster5_genes,c_6=cluster6_genes,
                  c_7=cluster7_genes,c_8=cluster8_genes,c_9=cluster9_genes,c_10=cluster10_genes,c_11=cluster11_genes,c_12=cluster12_genes)

for(df_name in names(df_list)) {
  df_list[[df_name]] = df_list[[df_name]] %>% select(donor, genes_list[[df_name]])
  df_list[[df_name]] = df_list[[df_name]] %>%
    rename_with(.fn = ~paste(., df_name, sep = "_"), .cols = c(genes_list[[df_name]]))
}

rm(c_0,c_1,c_2,c_3,c_4,c_5,c_6,c_7,c_8,c_9,c_10,c_11,c_12)

expressions_df = as.data.frame(df_list %>% reduce(left_join, by = "donor"))
```

```{r outcomes derive, include = FALSE}
# outcomes data
read_excel_allsheets <- function(filename, tibble = FALSE) {
    # I prefer straight data.frames
    # but if you like tidyverse tibbles (the default with read_excel)
    # then just pass tibble = TRUE
    sheets <- readxl::excel_sheets(filename)
    x <- lapply(sheets, function(X) readxl::read_excel(filename, sheet = X))
    if(!tibble) x <- lapply(x, as.data.frame)
    names(x) <- sheets
    x
}


a1ccpep = read_excel_allsheets("S:/Laura/BDC/Projects/Howard Davidson/R01/Data_Raw/Davidson pilot data/IDDA1C_Cpep.xlsx")
# remove all df as it looks incomplete
a1ccpep = a1ccpep[-1]

# change to numeric
a1c_df <- Reduce(function(x, y) merge(x, y, all = TRUE), a1ccpep)
a1c_df = a1c_df %>% filter(!is.na(`HWD ID`)) %>% mutate(a1c = as.numeric( case_when(A1c_Value == "NULL" ~ NA,
                                                                        TRUE ~ A1c_Value))) ; rm(a1ccpep)
# add in outcome vars
a1c_df = a1c_df %>% arrange(`HWD ID`, as.Date(VisitDate)) %>% group_by(`HWD ID`) %>% mutate(visit = row_number(),
                                                                                            a1c_g7 = ifelse(a1c > 7, "G7","L7")) # visit num
a1c_df = data.table(a1c_df, key = '`HWD ID`')
a1c_df = a1c_df[, days_from_v1 :=  cumsum(c(0, diff(as.Date(VisitDate)))), by=`HWD ID`]# days from baseline
a1c_df = a1c_df %>% mutate(y1_elig = ifelse(abs(days_from_v1- 365) < 90, 1, 0),
                           abs_y1_days = ifelse(abs(days_from_v1- 365) < 90, abs(days_from_v1- 365), NA))

a1c_y1 = a1c_df %>% group_by(`HWD ID`) %>% summarise(min_y1_days = min(abs_y1_days, na.rm = T))
a1c_y1 = a1c_df %>% group_by(`HWD ID`) %>% summarise(min_y1_days = min(abs_y1_days, na.rm = T))
a1c_y1$min_y1_days[is.infinite(a1c_y1$min_y1_days)] = NA
a1c_df = full_join(a1c_df, a1c_y1); rm(a1c_y1)
a1c_df = a1c_df %>% mutate(yr1a1c = ifelse(abs_y1_days == min_y1_days, a1c, NA))
a1c_df = a1c_df %>% group_by(`HWD ID`) %>% fill(yr1a1c)


avg_a1c = a1c_df %>% group_by(`HWD ID`) %>% summarise(meana1c = mean(a1c, na.rm = T),
                                                      firsta1c = first(a1c, na_rm = T),
                                                      maxa1c = max(a1c, na.rm=T),
                                                      yr1a1c = mean(yr1a1c, na.rm = T)) %>% mutate(id = `HWD ID`)
# add in slopes and auc
a1c_slopes = a1c_df %>% group_by(`HWD ID`) %>% do(tidy(lm(a1c ~ days_from_v1, data = .))) %>% filter(term == "days_from_v1") %>% select(`HWD ID`, estimate) %>% rename(a1c_slope = estimate)
avg_a1c = left_join(avg_a1c, a1c_slopes) ; rm(a1c_slopes)

avg_a1c$a1c_auc = sapply(split(a1c_df,a1c_df$`HWD ID`),function(df) trapz(df$days_from_v1,df$a1c))

# dichotomize
avg_a1c = avg_a1c %>% mutate(avg_a1c_g_med = ifelse(meana1c > median(meana1c,na.rm = T), "G_med", " L_med"),
                             first_a1c_g_med = ifelse(firsta1c > median(firsta1c,na.rm = T), "G_med", " L_med"),
                             y1_a1c_g_med = ifelse(yr1a1c > median(yr1a1c,na.rm = T), "G_med", " L_med"),
                             max_a1c_g_med = ifelse(maxa1c > median(maxa1c,na.rm = T), "G_med", " L_med"),
                             slope_a1c_g_med = ifelse(a1c_slope > median(a1c_slope,na.rm = T), "G_med", " L_med"),
                             auc_a1c_g_med = ifelse(a1c_auc > median(a1c_auc,na.rm = T), "G_med", " L_med"))


df2@meta.data = left_join(df2@meta.data, avg_a1c,  by=c('donor_number'='id'))
# View(df2@meta.data)

# add in other outcomes
# IDDA1c
avg_idda1c = a1c_df %>% group_by(`HWD ID`) %>% summarise(mean_idda1c = mean(IDDA1C, na.rm = T)) %>% mutate(id = `HWD ID`)
# add in slopes and auc
idda1c_slopes = a1c_df %>% group_by(`HWD ID`) %>% do(tidy(lm(IDDA1C ~ days_from_v1, data = .))) %>% filter(term == "days_from_v1") %>% 
  select(`HWD ID`, estimate) %>% rename(idda1c_slope = estimate)
avg_idda1c = left_join(avg_idda1c, idda1c_slopes) ; rm(idda1c_slopes)

avg_idda1c$idda1c_auc = sapply(split(a1c_df,a1c_df$`HWD ID`),function(df) trapz(df$days_from_v1,df$IDDA1C))

# dichotomize
avg_idda1c = avg_idda1c %>% mutate(avg_idda1c_g_med = ifelse(mean_idda1c > median(mean_idda1c,na.rm = T), "G_med", " L_med"),
                             slope_idda1c_g_med = ifelse(idda1c_slope > median(idda1c_slope,na.rm = T), "G_med", " L_med"),
                             auc_idda1c_g_med = ifelse(idda1c_auc > median(idda1c_auc,na.rm = T), "G_med", " L_med"))

df2@meta.data = left_join(df2@meta.data, avg_idda1c,  by=c('donor_number'='id'))


# total daily insulin dose
a1c_df$dose_udk = a1c_df$`U/day/kg`
avg_dose_udk = a1c_df %>% group_by(`HWD ID`) %>% summarise(mean_dose_udk = mean(dose_udk, na.rm = T)) %>% mutate(id = `HWD ID`)
# add in slopes and auc
dose_udk_slopes = a1c_df %>% group_by(`HWD ID`) %>% do(tidy(lm(dose_udk ~ days_from_v1, data = .))) %>% filter(term == "days_from_v1") %>% 
  select(`HWD ID`, estimate) %>% rename(dose_udk_slope = estimate)
avg_dose_udk = left_join(avg_dose_udk, dose_udk_slopes) ; rm(dose_udk_slopes)

avg_dose_udk$dose_udk_auc = sapply(split(a1c_df,a1c_df$`HWD ID`),function(df) trapz(df$days_from_v1,df$dose_udk))

# dichotomize
avg_dose_udk = avg_dose_udk %>% mutate(avg_dose_udk_g_med = ifelse(mean_dose_udk > median(mean_dose_udk,na.rm = T), "G_med", " L_med"),
                             slope_dose_udk_g_med = ifelse(dose_udk_slope > median(dose_udk_slope,na.rm = T), "G_med", " L_med"),
                             auc_dose_udk_g_med = ifelse(dose_udk_auc > median(dose_udk_auc,na.rm = T), "G_med", " L_med"))

df2@meta.data = left_join(df2@meta.data, avg_dose_udk,  by=c('donor_number'='id'))

# cpep
a1c_df = a1c_df %>% mutate(est_cpep = ifelse(`Est C-pep` > 0, `Est C-pep`, 0))
avg_est_cpep = a1c_df %>% group_by(`HWD ID`) %>% summarise(mean_est_cpep = mean(est_cpep, na.rm = T)) %>% mutate(id = `HWD ID`)
# add in slopes and auc
est_cpep_slopes = a1c_df %>% group_by(`HWD ID`) %>% do(tidy(lm(est_cpep ~ days_from_v1, data = .))) %>% filter(term == "days_from_v1") %>% 
  select(`HWD ID`, estimate) %>% rename(est_cpep_slope = estimate)
avg_est_cpep = left_join(avg_est_cpep, est_cpep_slopes) ; rm(est_cpep_slopes)

avg_est_cpep$est_cpep_auc = sapply(split(a1c_df,a1c_df$`HWD ID`),function(df) trapz(df$days_from_v1,df$est_cpep))

# dichotomize
avg_est_cpep = avg_est_cpep %>% mutate(avg_est_cpep_g_med = ifelse(mean_est_cpep > median(mean_est_cpep,na.rm = T), "G_med", " L_med"),
                             slope_est_cpep_g_med = ifelse(est_cpep_slope > median(est_cpep_slope,na.rm = T), "G_med", " L_med"),
                             auc_est_cpep_g_med = ifelse(est_cpep_auc > median(est_cpep_auc,na.rm = T), "G_med", " L_med"))
```

```{r variable selection and elnet, include = FALSE}
# A1c Slope

# donor and hwd id match
ids = df2@meta.data %>% select(donor, `HWD ID.y`) %>% unique()
colnames(ids) = c("donor", "HWD ID")
avg_a1c = full_join(avg_a1c, ids)

# merge with gene expressions
expressions_df$donor = factor(expressions_df$donor)
analysis_a1c = full_join(avg_a1c, expressions_df)
analysis_a1c_slope = analysis_a1c %>%  filter(!is.na(donor)) %>% select(a1c_slope, colnames(expressions_df))


#pare the list of 6000 genes down with univariate regression, p <.1
models_simple = list()
# with apply:
coefs_mat = expand.grid(c("a1c_slope"), colnames(expressions_df)[-1])

mods = apply(coefs_mat, 1, function(row) {
  lm(as.formula(paste(row[1], "~", row[2])), data = analysis_a1c_slope)
})
names(mods) = with(coefs_mat, paste(Var1, "vs", Var2))
coefs = lapply(mods, tidy, simplify = F)


# combine
uni_res = dplyr::bind_rows(coefs, .id = "mod")
uni_res = uni_res %>% filter(p.value < 0.1 & term != "(Intercept)")
new_genes = uni_res$term
#new_genes =  make.names(new_genes , unique = T, allow_ = F)


# scale and center the data
set.seed(1017)
analysis_a1c_slope = analysis_a1c_slope %>% select(-donor)
recipe_a1c_slope = recipe(analysis_a1c_slope) %>% update_role(everything()) %>% update_role(a1c_slope, new_role = "outcome")
recipe_a1c_slope = recipe_a1c_slope %>%
  step_zv(all_predictors()) %>%
  step_nzv(all_predictors()) %>%
  step_YeoJohnson(all_predictors()) %>%
  step_normalize(all_predictors()) %>%
  prep()
transformed_analysis_a1c_slope <- bake(recipe_a1c_slope,analysis_a1c_slope)
predictors_a1c_slope_en_large = colnames(transformed_analysis_a1c_slope)[-1]
predictors_a1c_slope_en = intersect(predictors_a1c_slope_en_large, new_genes)

# elnet small subset
# a1c_slope_en_large = easy_elasticnet(data = transformed_analysis_a1c_slope, outcome = "a1c_slope",cv_method = "loo",
#                               predictors = predictors_a1c_slope_en_large, model_type = "gaussian", out="1se.error")

# best model 0 covariates chosen; cvm = .0001124 best model with > 0 covariates selected 5776 features, cvm .0001128

a1c_slope_en_small = easy_elasticnet(data = transformed_analysis_a1c_slope, outcome = "a1c_slope",cv_method = "loo",
                              predictors = predictors_a1c_slope_en, model_type = "gaussian", out="1se.error")

```

# Methods

Inital dataframe was first subsetted by donor and cluster. Percent expressions were calculated by cluster such that each cluster was pared down to ~ 500 genes. This cutoff was between 20-30% by cluster (total genes n = 5815). Data were then scaled and centered. This set was further reduced via univariable linear regression with each outcome, with genes with p > 0.1 further removed (total genes n = 433). 

Leave One Out cross validation was performed for elasticnet variable selection. Models within 1 SE of the MSE of the best model were examined, with the best model being chosen by MSE. 

# A1c Slope Analysis

Elasticnets were performed on the large gene list (n = 5815 genes). The best models by Cross Validated Mean error selected 0 features from the list. The best model with > 0 features selected included 5776 features, but slightly higher CVMe. 

Elasticnets were then performed on the smaller gene list (n = 433 genes). The best model by CVMe included 61 genes listed below. Each suffix "_c_X" represents cluster.

```{r enet a1cslope}
print(a1c_slope_en_small)
```
