---
title: ""
author: "Tim Vigers & Laura Pyle"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_depth: 5
    toc_float: yes
---

```{r setup, include=FALSE}
library(arsenal)
library(tidyverse)
library(readxl)
library(performance)
library(knitr) 
library(data.table)
library(broom)
library(pROC)
library(caTools)

# analysis pckgs
library(Seurat) # need these for Differential expression testing
library(SeuratObject)
library(DESeq2)
#library(scran)
library(pzfx)
library(SingleCellExperiment)
library(S4Vectors)
# cell express %
library(scCustomize)

###
library(cowplot)
#library(Matrix.utils)
library(edgeR)
library(Matrix)
library(reshape2)
library(S4Vectors)
library(SingleCellExperiment)
library(pheatmap)
#library(apeglm)
library(png)
library(DESeq2)
library(RColorBrewer)
library(data.table)

###

# data
home_dir = "S:/Laura/BDC/Projects/Howard Davidson/R01/Data_Raw/Davidson pilot data/"
knitr::opts_knit$set(root.dir = home_dir)
#setwd("S:/Laura/BDC/Projects/Howard Davidson/R01/Data_Raw/Davidson pilot data/")
# cd4_processed = readRDS(paste0(home_dir, "OneDrive_1_6-28-2023/cd4_cells_processed.rds"))
# cd8_processed = readRDS(paste0(home_dir, "OneDrive_1_6-28-2023/cd8_cells_processed.rds"))
df = readRDS(paste0(home_dir, "OneDrive_1_6-28-2023/normalized.rds"))
#use regular clusters
df2 = df

# new clusters
#df2 = SetIdent(df, value= df@meta.data$RNA_clustifyr_celltype_individual)
#rm(df)
```

```{r, include = F}

```

```{r, include = FALSE}
# outcomes data
read_excel_allsheets <- function(filename, tibble = FALSE) {
    # I prefer straight data.frames
    # but if you like tidyverse tibbles (the default with read_excel)
    # then just pass tibble = TRUE
    sheets <- readxl::excel_sheets(filename)
    x <- lapply(sheets, function(X) readxl::read_excel(filename, sheet = X))
    if(!tibble) x <- lapply(x, as.data.frame)
    names(x) <- sheets
    x
}


a1ccpep = read_excel_allsheets("S:/Laura/BDC/Projects/Howard Davidson/R01/Data_Raw/Davidson pilot data/IDDA1C_Cpep.xlsx")
# remove all df as it looks incomplete
a1ccpep = a1ccpep[-1]

# change to numeric
a1c_df <- Reduce(function(x, y) merge(x, y, all = TRUE), a1ccpep)
a1c_df = a1c_df %>% filter(!is.na(`HWD ID`)) %>% mutate(a1c = as.numeric( case_when(A1c_Value == "NULL" ~ NA,
                                                                        TRUE ~ A1c_Value))) ; rm(a1ccpep)
# add in outcome vars
# We are interested in delta, auc, not so much mean 
a1c_df = a1c_df %>% arrange(`HWD ID`, as.Date(VisitDate)) %>% group_by(`HWD ID`) %>% mutate(visit = row_number()) # visit num
a1c_df = data.table(a1c_df, key = '`HWD ID`')

# delta
a1c_df_delta = a1c_df %>% filter(!is.na(a1c)) %>% group_by(`HWD ID`) %>%  mutate(duration_current_wk = as.numeric(difftime(as.Date(VisitDate), as.Date(`Date of Diagnosis`), units = "weeks")),
                                                   v_0y = ifelse(duration_current_wk == min(duration_current_wk), 1, 0),
                                                   v_2y = ifelse(abs(duration_current_wk-104) == min(abs(duration_current_wk - 104)), 1, 0))

# take the y0 and y2 dates and subtract a1c / time interval
delta_df = a1c_df_delta %>% filter(v_0y == 1 | v_2y == 1)
delta_df$v_2y = factor(delta_df$v_2y, labels = c("Y0", "Y2"))
delta_a1c = spread(delta_df%>% select(`HWD ID`, a1c, v_2y), v_2y, a1c)
delta_wks = spread(delta_df%>% select(`HWD ID`, duration_current_wk, v_2y), v_2y, duration_current_wk) %>% mutate(Y0_t = Y0, Y2_t = Y2) %>% select(`HWD ID`, Y0_t, Y2_t)
delta_df = full_join(delta_a1c, delta_wks); rm(delta_wks, delta_a1c, a1c_df_delta)
# calculate delta; this is curently in weeks so well need to divide by 52 to get it into terms of years
delta_df = delta_df %>% mutate(delta_a1c = (Y2- Y0)/(Y2_t/52 - Y0_t/52)) %>% select(`HWD ID`, delta_a1c)
a1c_df = left_join(a1c_df, delta_df)


a1c_df = a1c_df[, days_from_v1 :=  cumsum(c(0, diff(as.Date(VisitDate)))), by=`HWD ID`]# days from baseline

a1c_auc_df = a1c_df %>% group_by(`HWD ID`) %>% filter(!is.na(a1c)) %>% mutate(auc_trap = trapz(days_from_v1, a1c))
a1c_df = left_join(a1c_df, a1c_auc_df)
# fill in values
a1c_df = a1c_df %>% group_by(`HWD ID`) %>% fill(auc_trap)

a1c_df = a1c_df %>% mutate(id = `HWD ID`)

# dichotomize
a1c_df = a1c_df %>% mutate(auc_a1c_g_med = ifelse(auc_trap > median(auc_trap,na.rm = T), "G_med", " L_med"),
                           delta_a1c_g_med = ifelse(delta_a1c > median(delta_a1c,na.rm = T), "G_med", " L_med"))


# IDDA1c

# We are interested in delta, auc, not so much mean 
# delta
idda1c_df_delta = a1c_df %>% filter(!is.na(a1c)) %>% group_by(`HWD ID`) %>%  mutate(duration_current_wk = as.numeric(difftime(as.Date(VisitDate), as.Date(`Date of Diagnosis`), units = "weeks")),
                                                   v_0y = ifelse(duration_current_wk == min(duration_current_wk), 1, 0),
                                                   v_2y = ifelse(abs(duration_current_wk-104) == min(abs(duration_current_wk - 104)), 1, 0))

# take the y0 and y2 dates and subtract a1c / time interval
delta_df = idda1c_df_delta %>% filter(v_0y == 1 | v_2y == 1)
delta_df$v_2y = factor(delta_df$v_2y, labels = c("Y0", "Y2"))
delta_idda1c = spread(delta_df%>% select(`HWD ID`, IDDA1C, v_2y), v_2y, IDDA1C)

### CONFIRM THAT THIS IS DOING WHAT WE WANT FOR ALL SUBJ

delta_wks = spread(delta_df%>% select(`HWD ID`, duration_current_wk, v_2y), v_2y, duration_current_wk) %>% mutate(Y0_t = Y0, Y2_t = Y2) %>% select(`HWD ID`, Y0_t, Y2_t)
delta_df = full_join(delta_a1c, delta_wks); rm(delta_wks, delta_a1c, a1c_df_delta)
# calculate delta; this is curently in weeks so well need to divide by 52 to get it into terms of years
delta_df = delta_df %>% mutate(delta_a1c = (Y2- Y0)/(Y2_t/52 - Y0_t/52)) %>% select(`HWD ID`, delta_a1c)
a1c_df = left_join(a1c_df, delta_df)


a1c_df = a1c_df[, days_from_v1 :=  cumsum(c(0, diff(as.Date(VisitDate)))), by=`HWD ID`]# days from baseline

a1c_auc_df = a1c_df %>% group_by(`HWD ID`) %>% filter(!is.na(a1c)) %>% mutate(auc_trap = trapz(days_from_v1, a1c))
a1c_df = left_join(a1c_df, a1c_auc_df)
# fill in values
a1c_df = a1c_df %>% group_by(`HWD ID`) %>% fill(auc_trap)



df2@meta.data = left_join(df2@meta.data, avg_a1c,  by=c('donor_number'='id'))
# View(df2@meta.data)

# add in other outcomes
# IDDA1c
avg_idda1c = a1c_df %>% group_by(`HWD ID`) %>% summarise(mean_idda1c = mean(IDDA1C, na.rm = T)) %>% mutate(id = `HWD ID`)
# add in slopes and auc
idda1c_slopes = a1c_df %>% group_by(`HWD ID`) %>% do(tidy(lm(IDDA1C ~ days_from_v1, data = .))) %>% filter(term == "days_from_v1") %>% 
  select(`HWD ID`, estimate) %>% rename(idda1c_slope = estimate)
avg_idda1c = left_join(avg_idda1c, idda1c_slopes) ; rm(idda1c_slopes)

avg_idda1c$idda1c_auc = sapply(split(a1c_df,a1c_df$`HWD ID`),function(df) trapz(df$days_from_v1,df$IDDA1C))

# dichotomize
avg_idda1c = avg_idda1c %>% mutate(avg_idda1c_g_med = ifelse(mean_idda1c > median(mean_idda1c,na.rm = T), "G_med", " L_med"),
                             slope_idda1c_g_med = ifelse(idda1c_slope > median(idda1c_slope,na.rm = T), "G_med", " L_med"),
                             auc_idda1c_g_med = ifelse(idda1c_auc > median(idda1c_auc,na.rm = T), "G_med", " L_med"))

df2@meta.data = left_join(df2@meta.data, avg_idda1c,  by=c('donor_number'='id'))


# total daily insulin dose
a1c_df$dose_udk = a1c_df$`U/day/kg`
avg_dose_udk = a1c_df %>% group_by(`HWD ID`) %>% summarise(mean_dose_udk = mean(dose_udk, na.rm = T)) %>% mutate(id = `HWD ID`)
# add in slopes and auc
dose_udk_slopes = a1c_df %>% group_by(`HWD ID`) %>% do(tidy(lm(dose_udk ~ days_from_v1, data = .))) %>% filter(term == "days_from_v1") %>% 
  select(`HWD ID`, estimate) %>% rename(dose_udk_slope = estimate)
avg_dose_udk = left_join(avg_dose_udk, dose_udk_slopes) ; rm(dose_udk_slopes)

avg_dose_udk$dose_udk_auc = sapply(split(a1c_df,a1c_df$`HWD ID`),function(df) trapz(df$days_from_v1,df$dose_udk))

# dichotomize
avg_dose_udk = avg_dose_udk %>% mutate(avg_dose_udk_g_med = ifelse(mean_dose_udk > median(mean_dose_udk,na.rm = T), "G_med", " L_med"),
                             slope_dose_udk_g_med = ifelse(dose_udk_slope > median(dose_udk_slope,na.rm = T), "G_med", " L_med"),
                             auc_dose_udk_g_med = ifelse(dose_udk_auc > median(dose_udk_auc,na.rm = T), "G_med", " L_med"))

df2@meta.data = left_join(df2@meta.data, avg_dose_udk,  by=c('donor_number'='id'))

# cpep
a1c_df = a1c_df %>% mutate(est_cpep = ifelse(`Est C-pep` > 0, `Est C-pep`, 0))
avg_est_cpep = a1c_df %>% group_by(`HWD ID`) %>% summarise(mean_est_cpep = mean(est_cpep, na.rm = T)) %>% mutate(id = `HWD ID`)
# add in slopes and auc
est_cpep_slopes = a1c_df %>% group_by(`HWD ID`) %>% do(tidy(lm(est_cpep ~ days_from_v1, data = .))) %>% filter(term == "days_from_v1") %>% 
  select(`HWD ID`, estimate) %>% rename(est_cpep_slope = estimate)
avg_est_cpep = left_join(avg_est_cpep, est_cpep_slopes) ; rm(est_cpep_slopes)

avg_est_cpep$est_cpep_auc = sapply(split(a1c_df,a1c_df$`HWD ID`),function(df) trapz(df$days_from_v1,df$est_cpep))

# dichotomize
avg_est_cpep = avg_est_cpep %>% mutate(avg_est_cpep_g_med = ifelse(mean_est_cpep > median(mean_est_cpep,na.rm = T), "G_med", " L_med"),
                             slope_est_cpep_g_med = ifelse(est_cpep_slope > median(est_cpep_slope,na.rm = T), "G_med", " L_med"),
                             auc_est_cpep_g_med = ifelse(est_cpep_auc > median(est_cpep_auc,na.rm = T), "G_med", " L_med"))

df2@meta.data = left_join(df2@meta.data, avg_est_cpep,  by=c('donor_number'='id'))
```

```{r pseudobulk, include=FALSE}

# code modified from seurat website

# Extract raw counts and metadata to create SingleCellExperiment object
counts <- df2@assays$RNA@counts 

metadata <- df2@meta.data

# Set up metadata as desired for aggregation and DE analysis
metadata$cluster_id <- factor(df2@active.ident)

# Create single cell experiment object
sce <- SingleCellExperiment(assays = list(counts = counts), 
                           colData = metadata)

# Explore the raw counts for the dataset

## Check the assays present
# assays(sce)

## Check the counts matrix
# dim(counts(sce))
# counts(sce)[1:6, 1:6]

# # Explore the cellular metadata for the dataset
# 
# dim(colData(sce))
# head(colData(sce))

# Extract unique names of clusters (= levels of cluster_id factor variable)
cluster_names <- levels(colData(sce)$cluster_id)
# cluster_names
# 
# # Total number of clusters = 13
# length(cluster_names)

# Extract unique names of samples (= levels of sample_id factor variable)
 sample_names <- levels(colData(sce)$donor)
# sample_names
# 
# # Total number of samples
# length(sample_names)

# Subset metadata to include only the variables you want to aggregate across (here, we want to aggregate by sample and by cluster)
groups <- colData(sce)[, c("cluster_id", "donor")]
head(groups)

# Aggregate across cluster-sample groups
# transposing row/columns to have cell_ids as row names matching those of groups
aggregate.Matrix<-function(x,groupings=NULL,form=NULL,fun='sum',...)
{
  if(!is(x,'Matrix'))
    x<-Matrix(as.matrix(x),sparse=TRUE)
  if(fun=='count')
    x<-x!=0
  groupings2<-groupings
  if(!is(groupings2,'data.frame'))
    groupings2<-as(groupings2,'data.frame')
  groupings2<-data.frame(lapply(groupings2,as.factor))
  groupings2<-data.frame(interaction(groupings2,sep = '_'))
  colnames(groupings2)<-'A'
  if(is.null(form))
    form<-as.formula('~0+.')
  form<-as.formula(form)
  mapping<-dMcast(groupings2,form)
  colnames(mapping)<-substring(colnames(mapping),2)
  result<-t(mapping) %*% x
  if(fun=='mean')
    result@x<-result@x/(aggregate.Matrix(x,groupings2,fun='count'))@x
  attr(result,'crosswalk')<-grr::extract(groupings,match(rownames(result),groupings2$A))
  return(result)
}


aggr_counts <- aggregate.Matrix(t(counts(sce)), 
                                groupings = groups, fun = "sum") 

# Explore output matrix
# class(aggr_counts)
# dim(aggr_counts)
# aggr_counts[1:6, 1:6]
# Transpose aggregated matrix to have genes as rows and samples as columns
aggr_counts <- t(aggr_counts)
aggr_counts[1:6, 1:6]

```


```{r Findmarkers, include = FALSE}
#view(df2@meta.data)
############ A1C ##############
# set a1c groups as the ident
# avg a1c
Idents(df2) = df2$avg_a1c_g_med
a1c_avg_markers = FindMarkers(df2, ident.1 = 'G_med', ident.2 = ' L_med')

# slope a1c
Idents(df2) = df2$slope_a1c_g_med
a1c_slope_markers = FindMarkers(df2, ident.1 = 'G_med', ident.2 = ' L_med')

# auc a1c
Idents(df2) = df2$auc_a1c_g_med
a1c_auc_markers = FindMarkers(df2, ident.1 = 'G_med', ident.2 = ' L_med')
###############################

############ IDDA1C ##############
# set a1c groups as the ident
# avg idda1c
Idents(df2) = df2$avg_idda1c_g_med
idda1c_avg_markers = FindMarkers(df2, ident.1 = 'G_med', ident.2 = ' L_med')

# slope idda1c
Idents(df2) = df2$slope_idda1c_g_med
idda1c_slope_markers = FindMarkers(df2, ident.1 = 'G_med', ident.2 = ' L_med')

# auc idda1c
Idents(df2) = df2$auc_idda1c_g_med
idda1c_auc_markers = FindMarkers(df2, ident.1 = 'G_med', ident.2 = ' L_med')
###############################

############ Daily Dose ##############
# set a1c groups as the ident
# avg dose_udk
Idents(df2) = df2$avg_dose_udk_g_med
dose_udk_avg_markers = FindMarkers(df2, ident.1 = 'G_med', ident.2 = ' L_med')

# slope dose_udk
Idents(df2) = df2$slope_dose_udk_g_med
dose_udk_slope_markers = FindMarkers(df2, ident.1 = 'G_med', ident.2 = ' L_med')

# auc dose_udk
Idents(df2) = df2$auc_dose_udk_g_med
dose_udk_auc_markers = FindMarkers(df2, ident.1 = 'G_med', ident.2 = ' L_med')
###############################

############ cpep ##############
# set a1c groups as the ident
# avg est_cpep
Idents(df2) = df2$avg_est_cpep_g_med
est_cpep_avg_markers = FindMarkers(df2, ident.1 = 'G_med', ident.2 = ' L_med')

# slope est_cpep
Idents(df2) = df2$slope_est_cpep_g_med
est_cpep_slope_markers = FindMarkers(df2, ident.1 = 'G_med', ident.2 = ' L_med')

# auc est_cpep
Idents(df2) = df2$auc_est_cpep_g_med
est_cpep_auc_markers = FindMarkers(df2, ident.1 = 'G_med', ident.2 = ' L_med')
###############################
```

# Main Questions

# A1c

## cutoff values used

-   mean a1c: `r median(avg_a1c$meana1c, na.rm = T)`
-   slope: `r median(avg_a1c$a1c_slope, na.rm = T)`
-   auc: `r median(avg_a1c$a1c_auc, na.rm = T)`

### Mean

```{r}
kable(a1c_avg_markers)
```

### Slope

```{r}
kable(a1c_slope_markers)
```

### AUC

```{r}
kable(a1c_auc_markers)
```

### Intersection

```{r}
# All 3
print("All 3 Groups")
Reduce(intersect, list(
  rownames(a1c_slope_markers),
  rownames(a1c_avg_markers),  rownames(a1c_auc_markers)
))

print("Slope and Avg")
Reduce(intersect, list(
  rownames(a1c_slope_markers),
  rownames(a1c_avg_markers)
))


print("Slope and AUC")
Reduce(intersect, list(
  rownames(a1c_slope_markers), rownames(a1c_auc_markers)
))


print("Avg and AUC")
Reduce(intersect, list(
  rownames(a1c_avg_markers), rownames(a1c_auc_markers)
))
```


# IDDA1C

## cutoff values used

-   mean a1c: `r median(avg_idda1c$mean_idda1c, na.rm = T)`
-   slope: `r median(avg_idda1c$idda1c_slope, na.rm = T)`
-   auc: `r median(avg_idda1c$idda1c_auc, na.rm = T)`

### Mean

```{r}
kable(idda1c_avg_markers)
```

### Slope

```{r}
kable(idda1c_slope_markers)
```

### AUC

```{r}
kable(idda1c_auc_markers)
```

### Intersection

```{r}
# All 3
print("All 3 Groups")
Reduce(intersect, list(
  rownames(idda1c_slope_markers),
  rownames(idda1c_avg_markers),  rownames(idda1c_auc_markers)
))

print("Slope and Avg")
Reduce(intersect, list(
  rownames(idda1c_slope_markers),
  rownames(idda1c_avg_markers)
))


print("Slope and AUC")
Reduce(intersect, list(
  rownames(idda1c_slope_markers), rownames(idda1c_auc_markers)
))


print("Avg and AUC")
Reduce(intersect, list(
  rownames(idda1c_avg_markers), rownames(idda1c_auc_markers)
))
```

# dose_udk

## cutoff values used

-   mean a1c: `r median(avg_dose_udk$mean_dose_udk, na.rm = T)`
-   slope: `r median(avg_dose_udk$dose_udk_slope, na.rm = T)`
-   auc: `r median(avg_dose_udk$dose_udk_auc, na.rm = T)`

### Mean

```{r}
kable(dose_udk_avg_markers)
```

### Slope

```{r}
kable(dose_udk_slope_markers)
```

### AUC

```{r}
kable(dose_udk_auc_markers)
```

### Intersection

```{r}
# All 3
print("All 3 Groups")
Reduce(intersect, list(
  rownames(dose_udk_slope_markers),
  rownames(dose_udk_avg_markers),  rownames(dose_udk_auc_markers)
))

print("Slope and Avg")
Reduce(intersect, list(
  rownames(dose_udk_slope_markers),
  rownames(dose_udk_avg_markers)
))


print("Slope and AUC")
Reduce(intersect, list(
  rownames(dose_udk_slope_markers), rownames(dose_udk_auc_markers)
))


print("Avg and AUC")
Reduce(intersect, list(
  rownames(dose_udk_avg_markers), rownames(dose_udk_auc_markers)
))
```

# est_cpep

## cutoff values used

-   mean a1c: `r median(avg_est_cpep$mean_est_cpep, na.rm = T)`
-   slope: `r median(avg_est_cpep$est_cpep_slope, na.rm = T)`
-   auc: `r median(avg_est_cpep$est_cpep_auc, na.rm = T)`

### Mean

```{r}
kable(est_cpep_avg_markers)
```

### Slope

```{r}
kable(est_cpep_slope_markers)
```

### AUC

```{r}
kable(est_cpep_auc_markers)
```

### Intersection

```{r}
# All 3
print("All 3 Groups")
Reduce(intersect, list(
  rownames(est_cpep_slope_markers),
  rownames(est_cpep_avg_markers),  rownames(est_cpep_auc_markers)
))

print("Slope and Avg")
Reduce(intersect, list(
  rownames(est_cpep_slope_markers),
  rownames(est_cpep_avg_markers)
))


print("Slope and AUC")
Reduce(intersect, list(
  rownames(est_cpep_slope_markers), rownames(est_cpep_auc_markers)
))


print("Avg and AUC")
Reduce(intersect, list(
  rownames(est_cpep_avg_markers), rownames(est_cpep_auc_markers)
))
```
